### PR TITLE
[SQL] Wire up new schema creation DDL statement

### PR DESCRIPTION
**Pull request by akedin**
Simple `REGISTER EXTERNAL SCHEMA` parser logic.
Decided to do `REGISTER EXTERNAL` for now after discussion in the doc. Can change it to `CREATE FOREIGN` or something else if we want. 

Allows creating new schemas under `"beam"`. Currently doesn't have any actual remote schema providers.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by akedin** [comment]
R: @apilloud @amaliujia @XuMingmin @xumingming

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java [hunk 1]
```diff
@@ -149,7 +149,7 @@ private void unparseColumn(SqlWriter writer, Schema.Field column) {
   private Table toTable() {
     return Table.builder()
         .type(SqlDdlNodes.getString(type))
-        .name(name.getSimple())
+        .name(name.names.get(name.names.size() - 1))
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java]: What's this change for?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/parser/BeamDDLTest.java [hunk 2]
```diff
@@ -168,6 +174,27 @@ public void testParseDropTable() throws Exception {
...
+
+    connection.createStatement().execute("register external schema newSchema type test");
+    assertThat(
+        connection.getRootSchema().getSubSchema("beam").getSubSchemaNames(), contains("newSchema"));
+    SchemaPlus beamSchema = connection.getRootSchema().getSubSchema("beam");
+    SchemaPlus newSchema = beamSchema.getSubSchema("newSchema");
+
+    connection
+        .createStatement()
+        .execute("create table beam.newSchema.someTable (field1 int) type test");
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/parser/BeamDDLTest.java]: Why we need to create table from `newSchema` manually?

Does this example still work:
```
CREATE FOREIGN SCHEMA
  TYPE 'bigquery'
  LOCATION 'dataset_example'
  AS bq;

CREATE FOREIGN SCHEMA
  TYPE 'hcatalog'
  LOCATION 'hive-server:2341'
  AS hive;

SELECT * 
  FROM 
       bq.table_example_bq AS bq_table1
     JOIN
       hive.table_example_hive AS hive_table1
     ON
       bq_table1.some_field = hive_table1.some_other_field;
```

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java (hunk 1 above)
Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java]: before this PR tables could only have simple names. Now you can `create schema.subschema.tablename`, so this change gives you the `tablename`

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java]: I see. Thanks!

File: sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl [hunk 3]
```diff
@@ -329,4 +329,40 @@ SqlSetOptionBeam SqlSetOptionBeam(Span s, String scope) :
...
+    SqlNode properties = null;
+}
+{
+    <REGISTER> <EXTERNAL> <SCHEMA> {
+        s.add(this);
+    }
+
+    name = CompoundIdentifier()
+
+    <TYPE>
```
[the remark this answers is not in the document]
Comment (akedin) [reply sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl]: This PR doesn't address this part, there's no schema provider wired up, the statement currently directly creates a Beam schema in Calcite. When implemented, we will look up a schema provider based on `TYPE`.

And how to handle Table Type should depend on the specific schema provider after that. Some schema providers should be able to handle multiple table types (e.g. different table formats should work with Hive metastore if we have corresponding IOs), some probably can't (e.g. if schema provider is DB-specific, I can think of something like BigQuery dataset schema provider). Some may be read-only (e.g. if we didn't get to implementing the create table for it).

[the remark this answers is not in the document]
Comment (XuMingmin) [reply sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl]: Good to me ATM.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlRegisterExternalSchema.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.calcite.sql.SqlSpecialOperator;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.util.Pair;
+
+/** Parse tree for {@code REGISTER EXTERNAL SCHEMA} statement. */
+public class SqlRegisterExternalSchema extends SqlDdl implements SqlExecutableStatement {
+
+  private static final SqlOperator OPERATOR =
+      new SqlSpecialOperator("REGISTER EXTERNAL SCHEMA", SqlKind.OTHER_DDL);
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlRegisterExternalSchema.java]: nit: This seems like it is `CREATE_SCHEMA` type? Or is there something that blocks us from using that type?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlRegisterExternalSchema.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  private final SqlNode type;
+  private final SqlNode location;
+  private final SqlNode properties;
+
+  private SqlIdentifier name;
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlRegisterExternalSchema.java]: `final`?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java (hunk 1 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java]: Good catch! I think `pair.right` is the same thing already parsed.

**Pull request closed without merging**